### ISSUE TITLE
Allow ignoring errors using code comments

### ISSUE THREAD

**Issue by llucax** [jsh9/pydoclint#54]
Like most linters, it would be very useful being able to ignore a certain type of error for a particular function/class/module.

With documentation linting is harder to achieve, as annotating docstrings 

Darglint [allowed this](https://github.com/terrencepreilly/darglint#ignoring-errors-in-a-docstring) but by annotating the docstrings, but that is not a good option, as the annotations will show up then in the help, which is unhelpful and confusing to users.

The flexibility of the scheme `# noqa: <error> <argument>` is very useful though. For example, to only ignore the checking of one argument, one can do:

```py
def a_bound_function(self, arg1):
  """Do something interesting.

  Args:
    arg1: The first argument.

  # noqa: DAR101 arg1
  """
```

Or to ignore only one type of exception in `Raises`:

```py
def always_raises_exception(x):
    """Raise a zero division error or type error.o

    Args:
      x: The argument which could be a number or could not be.

    Raises:
      ZeroDivisionError: If x is a number.  # noqa: DAR402
      TypeError: If x is not a number.  # noqa: DAR402

    """
    x / 0
```

I suggest, if possible, adding the `# noqa:` comments directly in the code, for the above examples:

```py
def a_bound_function(self,
    arg1 # noqa: DAR101
):
  """Do something interesting.

  Args:
    arg1: The first argument.
  """

def always_raises_exception(x):  # noqa: DAR402 ZeroDivisionError TypeError
    """Raise a zero division error or type error.o

    Args:
      x: The argument which could be a number or could not be.

    Raises:
      ZeroDivisionError: If x is a number.
      TypeError: If x is not a number.

    """
    x / 0
```

Using the equivalent codes from `pydoclint`, of course.

**Comment by jsh9** [comment]
Hi @llucax ,

pydoclint can do function-level violation omission in the flake8 mode.

For example,

```python
def myFunc(  # noqa: DOC101, DOC103
        arg1: int,
        arg2: str,
) -> None:
    """
    Do something

    Parameters
    ----------
    arg1 : float
        Arg 1
    """
    pass
```

The `# noqa` comment needs to be put on the line where the function name (`myFunc`) is.

But `# noqa` only takes effect if you run it via `flake8 myFile.py`, not when you run it via `pydoclint myFile.py`.

I mentioned in [this part of README](https://github.com/jsh9/pydoclint/blob/397d6b280fe74c1c11f19b93605491e4f9fc4c4c/README.md?plain=1#L104) that "noqa" omission in the native mode may be added if there are demands for it.

----

That said, I don't think per-argument omission is feasible in pydoclint's current architecture.  This is because when pydoclint uses `ast` to parse a function, the whole function signature (such as `def funcName(arg1: int, arg2: str, ...) -> None`) is parsed together, regardless of any line breaks.

I also checked the parsed node, and I could not find any trace of `# noqa` comments when they are put after a particular argument.

If we want to parse per-argument `# noqa`, we might need to write a custom code parser.  And as a result, we may end up with as slow a speed as Darglint?

**Comment by llucax** [comment]
Hi @jsh9,

I see, thanks for the detailed explanation and sorry for not noticing the `flake8` mode, we are not using `flake8` for the moment and never dived into the `flake8` world yet, but maybe I'll do it in the future. I think it would also be very helpful to include an example about how to add `# noqa`, as there is no explicit mention to it, nor to the limitations on where should it be put.

I think [astroid](https://github.com/pylint-dev/astroid), the parser used by `pyint`, can also gather the information at this level, but `pylint` is also very slow, not sure if it's because of the parser or something else.

Is there any way to add "arguments" to the `noqa` in `flake8` mode, so at least we could select which exceptions to ignore? As I don't really want to stop checking for all exceptions, only a few ones that are raised by functions called inside the body.

As a very hacky solution I could also do:

```py
def always_raises_exception(x):
    """Raise a zero division error or type error.o

    Args:
      x: The argument which could be a number or could not be.

    Raises:
      ZeroDivisionError: If x is a number.  # noqa: DAR402
      TypeError: If x is not a number.  # noqa: DAR402

    """
    x / 0
    # pylint: disable=unreachable
    raise ZeroDivisionError()  # type:ignore
    raise TypeError()  # type: ignore
```

But it is quite nasty. I think being able to ignore certain type of exceptions in the check is super important, because as shown in the code above, even python operators can raise exceptions, what would be the alternative, wrap the division in a try/except block and create a new `ZeroDivisionError()`?

If I do this:

```py
    try:
        return x / 0
    except ZeroDivisionError:
        raise
```

Then `pylint` complains about having an `except` block that just re-raises, which is completely unnecessary, so it is a valid complain :)

Then I have to do this, which is even more verbose:

```py
    try:
        return 2 / divisor
    except ZeroDivisionError as exc:
        raise ZeroDivisionError("Cannot divide by zero.") from exc
```

And also adds unnecessary runtime costs, as I'm creating new objects, and makes the stack trace harder to read, as I have 2 `ZeroDivisionError`s for no reason.

I don't think this feature is complete without being able to ignore specific exceptions.

**Comment by jsh9** [comment]
If you'd like to ignore `DOC104` and `DOC201` in a function, you can simply do this:

```python
def myFunc(...):  # noqa: DOC104, DOC201
    ....
```

Is this the granularity that you are looking for?

**Comment by llucax** [comment]
> Is this the granularity that you are looking for?

No, I meant being able to ignore certain exception types for `DOC502`, so being able to say something like:

```py
def raises_exceptions(x): # noqa: DOC502(ZeroDivisionError, TypeError)
    """Raise a zero division error or type error.o

    Args:
      x: The argument which could be a number or could not be.

    Raises:
      ZeroDivisionError: If x is a number.
      TypeError: If x is not a number.
      WeirdException: blah.
    """
    if x:
        x / 0
    else:
        raise MyException()
```

So this should complain about `MyException` not being documented, about `WeirdException` not being raised and nothing else (so `ZeroDivisionError`, `TypeError` should not trigger an error because they are declared to be ignored).

I guess the broad disabling of the whole check is not that bad, but it would be very nice to still catch mistakes even if there are some exceptions that are raised indirectly.

**Comment by llucax** [comment]
> Also, I added a section in the documentation on how to ignore certain violation codes: https://github.com/jsh9/pydoclint/commit/8e0cccee875d91f3734923f40b0a6dfbc46965fe

Thanks!

**Comment by jsh9** [comment]
Hi @llucax , what you are looking for (`# noqa: DOC502(ZeroDivisionError, TypeError), DOC501(AssertError)`) may be doable, but it might really slow down pydoclint significantly.

I'll label this as "may tackle in the future" and close this issue for now.

But if you have any additional questions or comments, please feel free to add them below.  Of if you'd like to re-open it for another reason, please feel free to do so.

Thanks!

**Comment by llucax** [comment]
Fair enough.

**Comment by llucax** [comment]
Hi @jsh9, sorry to comment on this closed issue, but maybe this info will change the perception about it and maybe consider reopening.

I just tried using `flake8` to run `pydoclint` so I can use the ignore rules. Running with `flake8` is not really *Slightly slower* as stated in the readme:

Running on https://github.com/frequenz-floss/frequenz-sdk-python/:
```console
$ time flake8 --select=DOC src
...
real    0m10.562s
user    1m1.794s
sys     0m0.530s
$ time pydoclint src
...
real    0m0.189s
user    0m0.156s
sys     0m0.033s
```

**`flake8` is 2 orders of magnitude (100x) slower than running directly** :exclamation: :exploding_head: 

It is still an improvement over `darglint` for sure, but still this information might help considering reopening this issue (and maybe updating the README to be more accurate about the differences between running as a `flake8` plugin or not).

**Comment by jsh9** [comment]
Hi @llucax , I pulled frequenz-sdk-python (the latest on the main branch: https://github.com/frequenz-floss/frequenz-sdk-python/commit/1186f6ff7539158c0a91f8a6401bf5d30af2eae3) and tested it myself:

```
flake8 --select=DOC src  1.80s user 0.24s system 462% cpu 0.441 total
pydoclint src  0.11s user 0.02s system 96% cpu 0.132 total
```
(somehow I didn't get the nice formatting that you had.  I'm using macOS with zsh.)

There is still a big difference between running with flake8 and with native pydoclint, but somehow not as drastic as on your computer/environment. 

I'm curious what flake8 version do you have on your end?  Mine is 6.0.0.

**Comment by real-yfprojects** [comment]
> (somehow I didn't get the nice formatting that you had. I'm using macOS with zsh.)

I guess you are referring to the markdown formatting in his comment.

That can be done by writing
``````md
```console
$ command
output
```
``````
